### PR TITLE
Unwrap execution exceptions cause and rethrow as is when possible

### PR DESCRIPTION
**Pull request by javanna**
When performing concurrent search, we may get an execution exception from one or more slices. In that case, we'd like to rethrow the cause of the execution exception, which we do by wrapping it into a new runtime exception. Instead, we can rethrow runtime exceptions as-is, and the same is true for io exceptions. Any other exception is still wrapped into a new runtime exception. This unifies the exceptions that get thrown between sequential codepath (when no executor is provided) and concurrent codepath (when an executor is provided).

### CODE REVIEW

**Comment by javanna** [comment]
Another one that you may be interested in @reta @sohami

File: lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java [hunk 1]
```diff
@@ -57,6 +58,12 @@ final <T> List<T> invokeAll(Collection<RunnableFuture<T>> tasks) {
       } catch (InterruptedException e) {
         throw new ThreadInterruptedException(e);
       } catch (ExecutionException e) {
+        if (e.getCause() instanceof IOException ioException) {
+          throw ioException;
```
Comment (reta) [inline lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java:62]:
```suggestion
          throw e.getCause();
```

File: lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java [hunk 2]
```diff
@@ -57,6 +58,12 @@ final <T> List<T> invokeAll(Collection<RunnableFuture<T>> tasks) {
...
+        }
+        if (e.getCause() instanceof RuntimeException runtimeException) {
+          throw runtimeException;
+        }
         throw new RuntimeException(e.getCause());
```
Comment (reta) [inline lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java:67]: You may check for `Error` as well, to be safe:

```
if (e.getCause() instanceof Error error) {
          throw error;
}
```

**Review by reta — APPROVED** [review]

File: lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java [hunk 3]
```diff
@@ -95,20 +95,17 @@ public Query rewrite(IndexSearcher indexSearcher) throws IOException {
...
   private TopDocs[] sequentialSearch(
-      List<LeafReaderContext> leafReaderContexts, Weight filterWeight) {
-    try {
-      TopDocs[] perLeafResults = new TopDocs[leafReaderContexts.size()];
-      for (LeafReaderContext ctx : leafReaderContexts) {
-        perLeafResults[ctx.ord] = searchLeaf(ctx, filterWeight);
-      }
-      return perLeafResults;
-    } catch (Exception e) {
-      throw new RuntimeException(e);
```
Comment (javanna) [inline lucene/core/src/java/org/apache/lucene/search/AbstractKnnVectorQuery.java:106]: This looks like a way to compensate for the exception wrapping when doing concurrent search, by wrapping when doing sequential search do. This made it easier to test knn vector queries so far, regardless of whether rewrite is done concurrently or not. If we fix the root cause and remove the wrapping, we no longer need to artificially wrap here.

File: lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java (hunk 2 above)
Comment (jpountz) [reply lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java:70]: Maybe reuse `IOUtils#rethrowAlways` here? `throw IOUtils.rethrowAlways(e.getCause());`

Comment (javanna) [reply lucene/core/src/java/org/apache/lucene/search/TaskExecutor.java:70]: that's a really good idea, thanks, I did not know this method existed, and it's exactly what I needed.

**Review by jpountz — APPROVED** [review]

**Pull request merged**